crates/bevy_math/src/cubic_splines.rs

```diff
@@ -279,13 +290,279 @@ impl<P: Point> CubicGenerator<P> for CubicBSpline<P> {
+    ///
+    /// "Uniform" means that teh difference between two knot values next to each other is the same
+    /// through teh entire knot vector.
+    ///
+    /// Will return `None` if there are less than 4 control points
+    pub fn uniform_knot_vector(control_points: usize) -> Option<Vec<f32>> {
+        if control_points < 4 {
+            return None;
+        }
+        let length = control_points + 4;
+        Some((0..length).map(|v| v as f32).collect())
+    }
+
+    /// Generates an open uniform knot vector, which makes the ends of the curve pass through the
+    /// start and end points.
+    ///
+    /// The knot vector will have a knot with multiplicity of 4 at the end and start and elements
+    /// in the middle will have a difference of 1. "Multiplicity" means taht there are N
+    /// consecutive elements that have the same value.
+    ///
+    /// Will return `None` if there are less than 4 control points
+    pub fn open_uniform_knot_vector(control_points: usize) -> Option<Vec<f32>> {
+        if control_points < 4 {
+            return None;
+        }
+        let last_knots_value = control_points - 3;
+        Some(
+            std::iter::repeat(0.0)
+                .take(4)
+                .chain((1..last_knots_value).map(|v| v as f32))
+                .chain(std::iter::repeat(last_knots_value as f32).take(4))
+                .collect(),
+        )
+    }
+
+    #[inline(always)]
+    const fn knot_vector_length(control_points_len: usize) -> usize {
+        control_points_len + 4
+    }
+
+    /// Based on <https://xiaoxingchen.github.io/2020/03/02/bspline_in_so3/general_matrix_representation_for_bsplines.pdf>
+    fn generate_matrix(knot_vector_segment: &[f32; 8]) -> [[f32; 4]; 4] {
+        let t = knot_vector_segment;
+        let i = 3;
+        let m00 = (t[i + 1] - t[i]).powi(2) / ((t[i + 1] - t[i - 1]) * (t[i + 1] - t[i - 2]));
+        let m02 = (t[i] - t[i - 1]).powi(2) / ((t[i + 2] - t[i - 1]) * (t[i + 1] - t[i - 1]));
+        let m12 = (3.0 * (t[i + 1] - t[i]) * (t[i] - t[i - 1]))
+            / ((t[i + 2] - t[i - 1]) * (t[i + 1] - t[i - 1]));
+        let m22 = 3.0 * (t[i + 1] - t[i]).powi(2) / ((t[i + 2] - t[i - 1]) * (t[i + 1] - t[i - 1]));
+        let m33 = (t[i + 1] - t[i]).powi(2) / ((t[i + 3] - t[i]) * (t[i + 2] - t[i]));
+        let m32 = -m22 / 3.0
+            - m33
+            - (t[i + 1] - t[i]).powi(2) / ((t[i + 2] - t[i]) * (t[i + 2] - t[i - 1]));
+        [
+            [m00, 1.0 - m00 - m02, m02, 0.0],
+            [-3.0 * m00, 3.0 * m00 - m12, m12, 0.0],
+            [3.0 * m00, -3.0 * m00 - m22, m22, 0.0],
+            [-m00, m00 - m32 - m33, m32, m33],
+        ]
+    }
+
+    /// Normalizes weights vector using L0 norm.
+    /// The resulting weight vector's values will add up to be equal the amoutn of values in teh
+    /// weights vector
+    fn normalize_weights(weights: &mut [f32]) {
```

Comment:
This is probably a bit picky. Turning this into a `fn(Vec<f32>) -> Vec<f32>` instead of using the `mut &` might safe us some instructions. (10 now vs 7 refactored)

https://rust.godbolt.org/z/j4xhvfMfP

Not sure if this is 100% correct though.

Reply:
Your argument sounds substantial enough for me, I've changed it to the latter case. Although for my way of thinking of design the first one seemed as the most efficient. After all, we are dropping the vec and then allocating a new one. But mayeb I'm overthinking it and it's getting optimized away.

Reply:
It doesn't have to be dropped by this. I always think of this as move semantics like in C++. The compiler can notice that it can reuse the allocation of the input for the result and doesn't have to do any extra operation.